### PR TITLE
feat: cancel stale discover search requests with AbortController

### LINKED ISSUE #104: feat: cancel stale discover search requests with AbortController
## Context

**User Story:** As a user rapidly navigating back/forward on the discover page, I want search results to always reflect the current URL query so that stale results from a previous search never overwrite the current one.

When multiple searches are triggered in quick succession (e.g., rapid back/forward navigation causing URL `?q=` changes, or a future debounced live-search feature), the `handleSearch` function in `DiscoverContent` fires concurrent `fetch` requests with no cancellation mechanism. If the earlier request resolves after the later one, its stale results overwrite the current results — a classic race condition.

**Origin:** Identified during [PR #103 review](https://github.com/Chalet-Labs/contentgenie/pull/103#discussion_r2801042069) by Copilot. Deferred as a follow-up since the current search-on-submit pattern makes this low-probability, but it becomes critical if debounced live search is added later.

## Current Behavior

1. User navigates to `/discover?q=lex` (search fires for "lex")
2. User quickly navigates to `/discover?q=tech` (search fires for "tech")
3. If "lex" response arrives **after** "tech" response, "lex" results overwrite "tech" results
4. User sees results for "lex" while URL says `?q=tech`

## Proposed Solution

Use `AbortController` in the `useEffect` cleanup to cancel in-flight requests when a new search is initiated. When `urlQuery` changes, React runs cleanup for the previous effect, which aborts the stale request.

## Acceptance Criteria

- [ ] `VERIFY:` When a new search triggers while a previous fetch is in-flight, the previous fetch is aborted via `AbortController.abort()`
- [ ] `VERIFY:` Aborted fetches do not update component state (no stale results, no error flash)
- [ ] `VERIFY:` The `AbortError` from cancelled fetches is silently caught (not logged as an error, not shown to user)
- [ ] `VERIFY:` Normal search behavior is unchanged — single searches work exactly as before
- [ ] `VERIFY:` Existing tests pass (`bun run test`)
- [ ] `VERIFY:` New test covers: rapid sequential searches only show results from the latest query
- [ ] `VERIFY:` Lint passes (`bun run lint`)
- [ ] `VERIFY:` Build succeeds (`bun run build`)

## Edge Cases

- **AbortError handling:** `fetch` throws an `AbortError` when cancelled. This must be caught and silently ignored (not logged to console, not shown as a user-facing error).
- **Component unmount during fetch:** If the component unmounts while a fetch is in-flight, the abort cleanup prevents state updates on an unmounted component.
- **Empty query after non-empty:** Switching from `?q=lex` to no query should abort any in-flight search and clear results normally.

### Files to Modify

| File | Change |
|------|--------|
| `src/app/(app)/discover/discover-content.tsx` | Add AbortController to fetch, wire cleanup in useEffect |
| `src/app/(app)/discover/__tests__/discover-content.test.tsx` | Add test for stale response rejection |

### Approach

1. **Refactor `useEffect` to create an AbortController:**
   - Create a new `AbortController` at the start of the effect
   - Pass `controller.signal` to the `fetch()` call inside `handleSearch`
   - Return a cleanup function that calls `controller.abort()`
   - This may require inlining the fetch logic in the effect or threading the signal through `handleSearch`

2. **Handle AbortError in catch block:**
   - Check `if (err instanceof DOMException && err.name === 'AbortError') return;`
   - This prevents aborted fetches from setting error state or logging to console

3. **Add test for race condition:**
   - Mock two sequential fetches where the first resolves after the second
   - Assert that only the second fetch's results are displayed

### Patterns to Follow

The current `handleSearch` + `useEffect` pattern in `src/app/(app)/discover/discover-content.tsx` should be preserved structurally. Use the standard React pattern of AbortController in useEffect cleanup.

### Constraints

- Must not change user-facing behavior for normal (non-concurrent) searches
- Must not introduce a dependency — use the built-in `AbortController` Web API
- The `isLoading` state should only reflect the latest in-flight request

### In Scope
- Adding AbortController to cancel stale fetch requests in discover search
- Silently handling AbortError
- Test for race condition scenario

### Out of Scope
- Debounced live search (separate feature)
- Caching search results
- Request deduplication for identical queries
- Other fetch calls outside the discover page

## Dependencies

- Blocked by: PR #103 (must be merged first — this builds on the URL-based search)
- Related: #85 (parent issue for URL-based search)

## Testing Requirements

- [ ] Unit test: rapid sequential URL changes only show latest results
- [ ] Unit test: aborted fetch does not set error state
- [ ] Existing tests continue to pass unchanged

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Integrate `AbortController` into the discover page search `useEffect` to cancel in-flight fetch requests when the URL query parameter changes
- Prevent stale search results from overwriting current results during rapid back/forward navigation
- Silently catch `AbortError` exceptions (no error flash, no console noise) via `controller.signal.aborted` check
- Remove unnecessary `useCallback` wrapper — fetch logic inlined directly into `useEffect`

## Changes

| File | Description |
|------|-------------|
| `src/app/(app)/discover/discover-content.tsx` | AbortController per effect execution, signal passed to fetch, cleanup aborts stale requests |
| `src/app/(app)/discover/__tests__/discover-content.test.tsx` | 3 new tests: stale request cancellation, AbortError suppression, unmount cleanup |
| `CHANGELOG.md` | Entry under `[Unreleased] > Fixed` |

## Test plan

- [x] `bun run lint` — no warnings or errors
- [x] `bun run test` — 389/389 tests pass (10 discover-content tests, 3 new)
- [x] `bun run build` — production build succeeds
- [ ] Manual: search for "lex", quickly search for "huberman" — only huberman results appear
- [ ] Manual: search, navigate away before results — no console errors
- [ ] Manual: rapid back/forward between search queries — correct results for final query

Closes #104

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Stale search requests are now canceled when your query changes, preventing outdated results from overwriting current search results.
* **Tests**
  * Added tests to verify request cancellation, proper handling of aborted requests, and fetch abort on component unmount.

### CODE REVIEW

File: src/app/(app)/discover/discover-content.tsx [hunk 1]
```diff
@@ -19,43 +19,54 @@ export function DiscoverContent() {
...
-      );
+        if (!response.ok) {
+          const data = await response.json();
+          throw new Error(data.error || "Failed to search podcasts");
+        }
 
-      if (!response.ok) {
         const data = await response.json();
-        throw new Error(data.error || "Failed to search podcasts");
+        setPodcasts(data.podcasts || []);
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/(app)/discover/discover-content.tsx]: Fixed: Added abort signal check after `response.json()` and before `setPodcasts` to guard against stale results from completed-but-superseded requests.
